pkg/loop/internal/chain_reader_test.go

```diff
@@ -0,0 +1,235 @@
+		versionedBytes := &pb.VersionedBytes{
+			Version: unsupportedVer, // Unsupported version
+			Data:    []byte("abcdf"),
+		}
+
+		err := decodeVersionedBytes(&decodedData, versionedBytes)
+		if err == nil || err.Error() != expected.Error() {
+			t.Errorf("expected error: %s, but got: %v", expected, err)
+		}
+	})
+}
+
+func TestChainReaderClient(t *testing.T) {
+	RunChainReaderInterfaceTests(t, &interfaceTester{})
+
+	lis := bufconn.Listen(1024 * 1024)
+	s := grpc.NewServer()
+	es := &errorServer{}
+	pb.RegisterChainReaderServer(s, es)
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			panic(err)
+		}
+	}()
+	defer s.Stop()
+	conn := connFromLis(t, lis)
+	client := &chainReaderClient{grpc: pb.NewChainReaderClient(conn)}
+	ctx := context.Background()
+
+	errorTypes := []error{
+		types.ErrInvalidType,
+		types.ErrFieldNotFound,
+	}
+
+	for _, errorType := range errorTypes {
+		es.err = errorType
+		t.Run("GetLatestValue unwraps errors from server "+errorType.Error(), func(t *testing.T) {
+			err := client.GetLatestValue(ctx, types.BoundContract{}, "method", "anything", "anything")
+			assert.IsType(t, errorType, err)
+		})
+	}
+
+	// make sure that errors come from client directly
+	es.err = nil
+	invalidTypeErr := types.ErrInvalidType
+	t.Run("GetLatestValue returns error if type cannot be encoded in the wire format", func(t *testing.T) {
+		err := client.GetLatestValue(ctx, types.BoundContract{}, "method", &cannotEncode{}, &TestStruct{})
+		assert.NotNil(t, errors.As(err, &invalidTypeErr))
+	})
+}
+
+type interfaceTester struct {
+	lis    *bufconn.Listener
+	server *grpc.Server
+	conn   *grpc.ClientConn
+	fs     *fakeCodecServer
+}
+
+var encoder = makeEncoder()
+
+func makeEncoder() cbor.EncMode {
+	opts := cbor.CoreDetEncOptions()
+	opts.Sort = cbor.SortCanonical
+	e, _ := opts.EncMode()
+	return e
+}
+
+func (it *interfaceTester) SetLatestValue(ctx context.Context, t *testing.T, testStruct *TestStruct) types.BoundContract {
+	it.fs.SetLatestValue(testStruct)
+	return types.BoundContract{}
+}
+
+func (it *interfaceTester) GetPrimitiveContract(ctx context.Context, t *testing.T) types.BoundContract {
+	return types.BoundContract{}
+}
+
+func (it *interfaceTester) GetSliceContract(ctx context.Context, t *testing.T) types.BoundContract {
+	return types.BoundContract{}
+}
+
+func (it *interfaceTester) GetAccountBytes(_ int) []byte {
+	return []byte{1, 2, 3}
+}
+
+func (it *interfaceTester) Setup(t *testing.T) {
+	lis := bufconn.Listen(1024 * 1024)
+	it.lis = lis
+	it.fs = &fakeCodecServer{lock: &sync.Mutex{}}
+	s := grpc.NewServer()
+	pb.RegisterChainReaderServer(s, &chainReaderServer{impl: it.fs})
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			panic(err)
+		}
+	}()
+}
+
+func (it *interfaceTester) Teardown(t *testing.T) {
+	if it.server != nil {
+		it.server.Stop()
+	}
+
+	if it.conn != nil {
+		require.NoError(t, it.conn.Close())
+	}
+
+	it.lis = nil
+	it.server = nil
+	it.conn = nil
+}
```

Comment:
I really don't think that `TearDown` needs to be a first class part of this framework. Why don't we just register `t.Cleanup` from the `Setup`?
```suggestion
	}()
	t.Cleanup(func() {
		if it.server != nil {
			it.server.Stop()
		}
	
		if it.conn != nil {
			require.NoError(t, it.conn.Close())
		}
	
		it.lis = nil
		it.server = nil
		it.conn = nil
	})
}
```

Reply:
Changed